contributors/design-proposals/controller_history.md

```diff
@@ -0,0 +1,447 @@
+     and StatefulSetSpec respectively.
+- The generated type(s) for a controller is/are the type of the Object(s) 
+generated by the controller. 
+    - Pod is a generated type for the ReplicaSet, DaemonSet, and StatefulSet 
+    controllers.
+    - PersistentVolumeClaim is also a generated type for the StatefulSet 
+    controller.
+- The current state of a controller is the union of the states of its generated 
+Objects along with its status.
+    - For ReplicaSet, DaemonSet, and StatefulSet, the current state of the 
+    corresponding controllers can be derived from Pods they contain and the 
+    ReplicasSetStatus, DaemonSetStatus, and StatefulSetStatus objects 
+    respectively.
+- For all specification type Objects for controllers, the target state is the 
+set of fields in the Object that determine the state to which the controller 
+attempts to evolve the system.
+    - This may not necessarily be all fields of Object.
+- The target Object state is the subset of the target state necessary to create 
+a Objects of the generated type(s). 
+- To make the this concrete, for the StatefulSet controller:
+   - `.Spec.Template`, `.Spec.Replicas`, and `.Spec.VolumeClaims` determine the 
+   target state. The controller "wants" to create `.Spec.Replicas` Pods 
+   generated from `.Spec.Template` and `.Spec.VolumeClaims`.
+   - `.Spec.Template` and `.Spec.VolumeClaims` are the target Object state. This 
+   is enough information for the controller to generate Pods and corresponding 
+   PVCs.
+- If a version of the target Object state was used to generate an Object that 
+has not yet been deleted, we refer to the version, and any snapshots of the 
+version, as live.
+
+## API Objects
+
+Kubernetes controllers already persist their current and target states to the 
+API Server. In order to maintain a history of revisions to specification type 
+Objects, we only need to persist snapshots of the target Object states 
+contained in the specification type when they are revised.
+
+One approach would be to, for every specification type,  have a 
+corresponding History type. For example, we could introduce a StatefulSetHistory 
+object that aggregates a PodTemplateSpec and a slice of PersistentVolumeClaims.
+The StatefulSet controller could use this object to store point in time 
+snapshots of versions of StatefulSetSpecs. However, this requires that we 
+introduce a new History Kind for all current and future controllers. It has the 
+benefit of type safety, but, for this benefit, we trade generality.
+
+Another approach would be to use PodTemplate objects. This mechanisms provides 
+the desired generality, but it only provides for the recording of versions of
+PodTemplateSpecs (e.g. For StatefulSet, we can not use PodTemplates to 
+record revisions to PersistentVolumeClaims). Also, it introduces the potential 
+for overlapping histories for two Objects of different Kinds, with the same 
```

Comment:
@kargakis Any collision will result in an increase in the uniquifier. Comparing the hash in the CR names will not work. We could label the CR's with the non-unique hash value if we want to optimize comparison though.